### ISSUE TITLE
Correct inner type annotation causes compiler error complaining that equal signatures are not matching

### ISSUE THREAD

**Issue by dennda** [elm-lang/elm-compiler#1301]
Hello! First of all, thanks for making elm!

Second, here's a fun error:
```
Warning: -- TYPE MISMATCH -------------------------------------------------- ././Tree.elm

The type annotation for `mapWithID` does not match its definition.

53│ mapWithID : ((NodeID, a) -> b) -> Tree a -> Tree b
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The type annotation is saying:

    (( NodeID, a ) -> b) -> Tree a -> Tree b

But I am inferring that the definition has this type:

    (( NodeID, a ) -> b) -> Tree a -> Tree b

Hint: A type annotation is clashing with itself or with a sub-annotation. This
can be particularly tricky, so read more about it.
<https://github.com/elm-lang/elm-compiler/blob/0.16.0/hints/type-annotations.md> Use --force to continue.
```

Note that it's complaining that the type annotation doesn't match… an equal definition.

You get that with this code snippet (which is incomplete / incorrect, but that's not the point):
```
mapWithID : ((NodeID, a) -> b) -> Tree a -> Tree b
mapWithID f tree =
  let
    nodeID = getNodeID tree
    value = getRootValue tree

    childrenWithID : List (NodeID, Tree a)
    childrenWithID = getChildren tree
      |> List.map (\child -> (getNodeID child, child))
  in
    Node nodeID (f (nodeID, value)) []
```

If you remove the `childrenWithID` annotation within `mapWithID` (which I believe to be correct), it will build just fine.

Complete code:
```
module Tree where


type alias NodeID = Int

type Tree a =
    Node NodeID a (List (Tree a))


-- Get the ID of the root node of a given tree.
getNodeID : Tree a -> NodeID
getNodeID (Node i _ _) = i


-- Get the children of the root node of a given tree.
getChildren : Tree a -> List (Tree a)
getChildren (Node _ _ c) = c


-- Get the value of a root node of a given tree.
getRootValue : Tree a -> a
getRootValue (Node _ v _) = v


-- Given a tree and a node ID, return the node with that ID if it exists.
getNodeWithID : NodeID -> Tree a -> Maybe (Tree a)
getNodeWithID nodeID tree =
  let
    currentNodeID = getNodeID tree
    children = getChildren tree
  in
    if currentNodeID == nodeID
    then
      Just tree
    else
      List.filterMap (getNodeWithID nodeID) children
        |> List.head


-- Incomplete/incorrect implementation, but that is not the point
mapWithID : ((NodeID, a) -> b) -> Tree a -> Tree b
mapWithID f tree =
  let
    nodeID = getNodeID tree
    value = getRootValue tree

    childrenWithID : List (NodeID, Tree a)
    childrenWithID = getChildren tree
      |> List.map (\child -> (getNodeID child, child))
  in
    Node nodeID (f (nodeID, value)) []
```

Thanks!

**Comment by jvoigtlaender** [comment]
The thing is that your "inner type annotation" is *not* correct from Elm's perspective. The `a` in `childrenWithID : List (NodeID, Tree a)` there is treated as a new `a'`, not as the `a` from `mapWithID : ((NodeID, a) -> b) -> Tree a -> Tree b`. You can read more about this in [this issue](https://github.com/elm-lang/elm-compiler/issues/1214), but note that in @evancz's first comment there the "can write" should actually be "can't write".

About the "complaining that the type annotation doesn't match… an equal definition", yeah, that happens currently, and there are other issues open about this already.

In summary, I think this issue here can be closed.

**Comment by evancz** [comment]
Going with @jvoigtlaender's analysis. Thanks @dennda for the report, and sorry for the trouble!

**Comment by dennda** [comment]
No trouble at all. Thank *you*! :-)

—
Enviado desde mi movíl
